### PR TITLE
add ldap-acl plugin to integrate acl and ldap-auth

### LINKED ISSUE #1439: Can't set ACL rules for LDAP-AUTH
### Summary

I want to control kong admin apis under `/admin` path, and control is by LDAP-AUTH plugin, but I found ACL plugin can not control credentials from LDAP-AUTH plugin, is there any way to implement this feature?

### Additional Details & Logs

- Kong version (0.8.3)

### PR DESCRIPTION
**Pull request by tonicospinelli**
### Summary

This `ldap-acl` plugin is a glue between `ldap-auth` and `acl` plugins.
It will find `consumer_id` by ldap username and append it on `ngx.ctx.authenticated_credential`, because `ldap-auth` plugin doesn't have a link with kong consumers.
#### Requirements
- sync all ldap users into kong consumers
- enable and configure `ldap-auth` and `acl` plugins
### Full changelog
- add `ldap-acl` makes `ldap-auth` and `acl` working togheter
### Issues resolved

Fix #1439

### CODE REVIEW

File: kong/plugins/ldap-acl/bind_consumer.lua [hunk 1]
```diff
@@ -0,0 +1,22 @@
+local BindConsumer = { dao = nil }
+
+function BindConsumer:new(o)
+  o = o or {} -- create object if user does not provide one
+  if not o.dao then
+    error("a dao is required")
+  end
```
Comment (Tieske) [inline kong/plugins/ldap-acl/bind_consumer.lua]: is the above check necesary? not checking it will also result in a 500 anyway.

File: kong/plugins/ldap-acl/bind_consumer.lua [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  return o
+end
+
+function BindConsumer.bind(self, username)
+  local consumers = self.dao:find_all { username = username }
+  if not consumers then
+    return nil
+  end
+  local consumer = consumers[1]
+  return consumer
```
Comment (Tieske) [inline kong/plugins/ldap-acl/bind_consumer.lua]: just do a final line; `if consumers then return consumer[1] end`

File: kong/plugins/ldap-acl/handler.lua [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  LdapAclHandler.super.access(self)
+
+  local ldap_authorization = ngx.ctx.authenticated_credential
+
+  if not ldap_authorization then
+    return responses.send_HTTP_UNAUTHORIZED()
+  end
+
+  if not ldap_authorization.username then
+    return responses.send_HTTP_UNAUTHORIZED("Unkown username")
```
Comment (Tieske) [inline kong/plugins/ldap-acl/handler.lua]: you're providing info to a potential attacker. Just respond, don't provide the reason why. Check the code, there are more of those.

File: kong/plugins/ldap-acl/handler.lua [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  end
+
+  local bind_consumer = BindConsumer:new { dao = singletons.dao.consumers, cache = cache }
+
+  ngx.log(ngx.DEBUG, "[ldap-acl] bind username between ldap and consumer:" .. ldap_authorization.username)
```
Comment (Tieske) [inline kong/plugins/ldap-acl/handler.lua]: do not use the `..` operator, just use comma's. `ngx.log`  will concatenate anyway. (arguments are evaluated before the function is called, so even if nothing is logged due to the level, the concatenation is still executed. If you pass the arguments, the log function will only concatenate when it actually logs). Check code for more of those

File: kong/plugins/ldap-acl/handler.lua (hunk 3 above)
Comment (tonicospinelli) [inline kong/plugins/ldap-acl/handler.lua]: it makes sense!! I will remove all

File: kong/plugins/ldap-acl/handler.lua [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+  local consumer = BindConsumer.bind(bind_consumer, ldap_authorization.username)
+
+  if not consumer then
+    ngx.log(ngx.ERR, "[ldap-acl] consumer was not found by ldap user(" .. ldap_authorization.username .. ")")
+    return responses.send_HTTP_UNAUTHORIZED("Unkown consumer")
+  end
+
+  ngx.ctx.authenticated_credential.consumer_id = consumer.id
```
Comment (Tieske) [inline kong/plugins/ldap-acl/handler.lua]: you're having a local `ldap_authorization` already, why not set that one instead of an expensive `ngx.ctx` table lookup

File: spec/03-plugins/17-ldap-acl/01-bind_consumer_spec.lua [hunk 6]
```diff
@@ -0,0 +1,44 @@
+local helpers = require "spec.helpers"
+
+describe('Busted unit testing framework', function()
```
Comment (Tieske) [inline spec/03-plugins/17-ldap-acl/01-bind_consumer_spec.lua]: thats' not useful information, we know that already

**Comment by Tieske** [comment]
I just had a quick glance over the code. Haven't thought about the use case etc, that's more up to @thibaultcha and @thefosk anyway.

also a general remark; it only has 2 unit tests, though are quite some edge cases to test, and integration is missing too...

**Comment by Tieske** [comment]
btw: did I mention that we like contributions? 😄   so thanks for that 👍

File: kong/plugins/ldap-acl/handler.lua (hunk 4 above)
Comment (tonicospinelli) [inline kong/plugins/ldap-acl/handler.lua]: thanks for the tip... I'm not a lua expert 😄

**Comment by tonicospinelli** [comment]
@Tieske very thanks for review! We started a POC using kong with ldap + acl and after that this PR was born.

Sorry, but my lua skills has 2 days only :smile: 

I'll create integration tests for it.

File: kong/plugins/ldap-acl/handler.lua (hunk 4 above)
Comment (Tieske) [inline kong/plugins/ldap-acl/handler.lua]: :+1:

**Comment by tonicospinelli** [comment]
@Tieske I did changes and now I only "owe you" the integration tests

**Comment by p0pr0ck5** [comment]
Hi @tonicospinelli,

Thanks for your PR. We will not be accepting community plugins in the core anymore. Feel free to publish this plugin on your own following the [distributing your plugin](https://getkong.org/docs/0.10.x/plugin-development/distribution/#distribute-your-plugin) guide. You are also encouraged to advertise it on the [mailing list](https://groups.google.com/forum/#!forum/konglayer), and we will soon provide a "community plugins" section on [getkong.org/plugins/](http://getkong.org/plugins/) on which you will be able to document it.

In the future, we plan to take out of this core repository all of the plugins we currently have as well.

Thank you!
